### PR TITLE
improve dataloader signals and messages

### PR DESCRIPTION
**Pull request by zhreshold**
## Description ##
Improve dataloader use experience.
With this PR, dataloaders are

- More responsive to terminate with Ctrl + C
- Show helpful messages if exceptions raises in multiple workers. DataLoaders used to produce unreadable messages.
- Less likely to hang forever, given timeout is added to fetch logic

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by zhreshold** [comment]
@szha, @eric-haibin-lin @sxjscience for review

**Comment by zhreshold** [comment]
ping for review 😄

File: python/mxnet/gluon/data/dataloader.py [hunk 1]
```diff
@@ -537,16 +557,21 @@ def default_batchify_fn(data):
...
     def __init__(self, dataset, batch_size=None, shuffle=False, sampler=None,
                  last_batch=None, batch_sampler=None, batchify_fn=None,
                  num_workers=0, pin_memory=False, pin_device_id=0,
-                 prefetch=None, thread_pool=False):
+                 prefetch=None, thread_pool=False, timeout=120):
         self._dataset = dataset
         self._pin_memory = pin_memory
         self._pin_device_id = pin_device_id
         self._thread_pool = thread_pool
+        self._timeout = timeout
```
Comment (eric-haibin-lin) [inline python/mxnet/gluon/data/dataloader.py]: add a check for positive values for timeout? Otherwise lgtm

**Comment by leezu** [comment]
I don't think it's a common or intended use-case to have workers process a batch for more than 120 seconds. It's still a breaking change I guess, but it may be fine to break a feature that's potentially unused (ie noone may rely on `timeout > 120`).

**Comment by wkcn** [comment]
I have a suggestion: Dataloader does not terminate the program but print a warning when timeout. Users decide whether to terminate it.

**Comment by zhreshold** [comment]
@wkcn It's mandatory to have timeout in order to catch excetions in subprocess due to a python bug.

**Pull request merged**